### PR TITLE
use sync id when recovering

### PR DESCRIPTION
**Pull request by brwe**
Skip phase 1 of recovery in case an identical sync id was found on primary and replica. Relates to #10032

### CODE REVIEW

File: src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 1]
```diff
@@ -153,7 +155,13 @@ public InternalEngine(EngineConfig engineConfig, boolean skipInitialTranslogReco
                 if (translogId.v1() != null && skipInitialTranslogRecovery == false) {
                     recoverFromTranslog(translogId.v1(), transformer);
                 } else {
-                    flush(true, true);
+                    if (lastCommittedSegmentInfos.getUserData().get(SYNC_COMMIT_ID) == null) {
```
Comment (bleskes) [inline src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: can we add a nocommit here plus a comment as to why we do this? I think we need to think about this some more but don't want to stop development now..

File: src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 2]
```diff
@@ -1263,9 +1274,15 @@ protected void recoverFromTranslog(long translogId, TranslogRecoveryPerformer ha
             IOUtils.closeWhileHandlingException(translog);
             throw new EngineException(shardId, "failed to recover from translog", e);
         }
-        flush(true, true);
+
         if (operationsRecovered > 0) {
+            flush(true, true);
             refresh("translog recovery");
+        } else if (lastCommittedSegmentInfos.getUserData().get(SYNC_COMMIT_ID) == null) {
```
Comment (bleskes) [inline src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: same comment about no commit.

File: src/main/java/org/elasticsearch/index/snapshots/blobstore/BlobStoreIndexShardRepository.java [hunk 3]
```diff
@@ -186,7 +186,8 @@ public IndexShardSnapshotStatus snapshotStatus(SnapshotId snapshotId, ShardId sh
 
     @Override
     public void verify(String seed) {
-        BlobContainer testBlobContainer = blobStore.blobContainer(basePath);;
+        BlobContainer testBlobContainer = blobStore.blobContainer(basePath);
+        ;
```
Comment (bleskes) [inline src/main/java/org/elasticsearch/index/snapshots/blobstore/BlobStoreIndexShardRepository.java]: extra semi-colon (was there before)

File: src/main/java/org/elasticsearch/index/store/Store.java [hunk 4]
```diff
@@ -665,24 +666,30 @@ public String toString() {
...
+            metadata = buildMetadata(commit, directory, logger, segmentCommitInfos);
             assert metadata.isEmpty() || numSegmentFiles() == 1 : "numSegmentFiles: " + numSegmentFiles();
         }
 
-        ImmutableMap<String, StoreFileMetaData> buildMetadata(IndexCommit commit, Directory directory, ESLogger logger) throws IOException {
+        ImmutableMap<String, StoreFileMetaData> buildMetadata(IndexCommit commit, Directory directory, ESLogger logger, SegmentInfos segmentCommitInfos) throws IOException {
             ImmutableMap.Builder<String, StoreFileMetaData> builder = ImmutableMap.builder();
             Map<String, String> checksumMap = readLegacyChecksums(directory).v1();
             try {
-                final SegmentInfos segmentCommitInfos = Store.readSegmentsInfo(commit, directory);
```
Comment (bleskes) [inline src/main/java/org/elasticsearch/index/store/Store.java]: we can't take this one out of the try catch, it protects agains corruption. I think we should rename `buildMetadata` to `loadMetadata` and set both commutUserData and metadata in it.

File: src/main/java/org/elasticsearch/index/store/Store.java [hunk 5]
```diff
@@ -969,9 +977,15 @@ public void readFrom(StreamInput in) throws IOException {
             for (int i = 0; i < size; i++) {
                 StoreFileMetaData meta = StoreFileMetaData.readStoreFileMetaData(in);
                 builder.put(meta.name(), meta);
+                logger.info("read {} {}", meta.name(), meta);
```
Comment (bleskes) [inline src/main/java/org/elasticsearch/index/store/Store.java]: left over debugging tool?

File: src/main/java/org/elasticsearch/index/store/Store.java [hunk 6]
```diff
@@ -969,9 +977,15 @@ public void readFrom(StreamInput in) throws IOException {
...
             }
             this.metadata = builder.build();
             assert metadata.isEmpty() || numSegmentFiles() == 1 : "numSegmentFiles: " + numSegmentFiles();
+            commitUserData = new HashMap<>();
+            int num = in.readVInt();
+            for (int i = num; i > 0; i--) {
+                commitUserData.put(in.readString(), in.readOptionalString());
```
Comment (bleskes) [inline src/main/java/org/elasticsearch/index/store/Store.java]: we don't need the optional string here as we found out :)

File: src/main/java/org/elasticsearch/index/store/Store.java [hunk 7]
```diff
@@ -980,6 +994,15 @@ public void writeTo(StreamOutput out) throws IOException {
             for (StoreFileMetaData meta : this) {
                 meta.writeTo(out);
             }
+            out.writeVInt(commitUserData.size());
+            for (Map.Entry<String, String> entry : commitUserData.entrySet()) {
+                out.writeString(entry.getKey());
+                out.writeOptionalString(entry.getValue());
```
Comment (bleskes) [inline src/main/java/org/elasticsearch/index/store/Store.java]: same here - no optional

File: src/main/java/org/elasticsearch/indices/recovery/RecoveryState.java [hunk 8]
```diff
@@ -811,15 +812,7 @@ public synchronized long totalRecoverBytes() {
             return total;
         }
 
-        public synchronized long totalReuseBytes() {
```
Comment (bleskes) [inline src/main/java/org/elasticsearch/indices/recovery/RecoveryState.java]: the fact that this is not reused is actually a bug in the toXContent:

```
builder.byteSizeField(Fields.TOTAL_IN_BYTES, Fields.TOTAL, totalBytes());
 builder.byteSizeField(Fields.REUSED_IN_BYTES, Fields.REUSED, totalBytes());
            
```

can we keep this here? I'll make a pull request..

File: src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java [hunk 9]
```diff
@@ -168,21 +170,33 @@ public void phase1(final SnapshotIndexCommit snapshot) throws ElasticsearchExcep
             // Generate a "diff" of all the identical, different, and missing
             // segment files on the target node, using the existing files on
             // the source node
-            final Store.RecoveryDiff diff = recoverySourceMetadata.recoveryDiff(new Store.MetadataSnapshot(request.existingFiles()));
+            final Store.RecoveryDiff diff;
+            boolean copyFiles = true;
+            if (recoverySourceMetadata.getCommitUserData().get(Engine.SYNC_COMMIT_ID) != null &&
```
Comment (bleskes) [inline src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: don't reuse a copyFile flag here as discussed.

File: src/main/java/org/elasticsearch/indices/recovery/StartRecoveryRequest.java [hunk 10]
```diff
@@ -46,7 +47,7 @@
 
     private boolean markAsRelocated;
 
-    private Map<String, StoreFileMetaData> existingFiles;
+    Store.MetadataSnapshot metadataSnapshot;
```
Comment (bleskes) [inline src/main/java/org/elasticsearch/indices/recovery/StartRecoveryRequest.java]: we lost the private?

File: src/main/java/org/elasticsearch/indices/recovery/StartRecoveryRequest.java [hunk 11]
```diff
@@ -126,10 +118,11 @@ public void writeTo(StreamOutput out) throws IOException {
...
         out.writeBoolean(markAsRelocated);
-        out.writeVInt(existingFiles.size());
-        for (StoreFileMetaData md : existingFiles.values()) {
-            md.writeTo(out);
-        }
+        metadataSnapshot.writeTo(out);
         out.writeByte(recoveryType.id());
     }
+
+    public Store.MetadataSnapshot metadataSnapshot() {
```
Comment (bleskes) [inline src/main/java/org/elasticsearch/indices/recovery/StartRecoveryRequest.java]: can we put this next to the other getters?

File: src/test/java/org/elasticsearch/gateway/RecoveryFromGatewayTests.java [hunk 12]
```diff
@@ -435,6 +439,65 @@ public void testReusePeerRecovery() throws Exception {
 
     @Test
     @Slow
+    @TestLogging("indices.recovery:TRACE,index.store:TRACE")
+    public void testSyncFlushedRecovery() throws Exception {
```
Comment (bleskes) [inline src/test/java/org/elasticsearch/gateway/RecoveryFromGatewayTests.java]: can we merge this test with the previous one where we randomly replace the first full cluster restart with an synced flush...

File: src/test/java/org/elasticsearch/index/store/StoreTest.java [hunk 13]
```diff
@@ -713,9 +713,9 @@ public void testRecoveryDiffWithLegacyCommit() {
         Map<String, StoreFileMetaData> metaDataMap = new HashMap<>();
         metaDataMap.put("segments_1", new StoreFileMetaData("segments_1", 50, null, null, new BytesRef(new byte[] {1})));
         metaDataMap.put("_0_1.del", new StoreFileMetaData("_0_1.del", 42, "foobarbaz", null, new BytesRef()));
-        Store.MetadataSnapshot first = new Store.MetadataSnapshot(metaDataMap);
+        Store.MetadataSnapshot first = new Store.MetadataSnapshot(metaDataMap, Collections.EMPTY_MAP);
```
Comment (bleskes) [inline src/test/java/org/elasticsearch/index/store/StoreTest.java]: can we add a test somewhere here (or extend an existing one) to see we read commit data correctly?

File: src/test/java/org/elasticsearch/indices/recovery/StartRecoveryRequestTest.java [hunk 14]
```diff
@@ -69,11 +73,42 @@ public void testSerialization() throws Exception {
...
-        assertThat(outRequest.existingFiles(), equalTo(inRequest.existingFiles()));
+        assertThat(outRequest.metadataSnapshot().asMap(), equalTo(inRequest.metadataSnapshot().asMap()));
         assertThat(outRequest.recoveryId(), equalTo(inRequest.recoveryId()));
         assertThat(outRequest.recoveryType(), equalTo(inRequest.recoveryType()));
     }
 
 
+    // TODO: where should this test be?
+    @Test
+    public void testMetadataSnapshotStreaming() throws Exception {
```
Comment (bleskes) [inline src/test/java/org/elasticsearch/indices/recovery/StartRecoveryRequestTest.java]: in StoreTest.java?

**Comment by bleskes** [comment]
looks good. left some comments as discussed.

File: src/test/java/org/elasticsearch/index/store/StoreTest.java (hunk 13 above)
Comment (brwe) [inline src/test/java/org/elasticsearch/index/store/StoreTest.java]: I added StoreTest.testUserDataRead(). Is this what you meant?

**Comment by brwe** [comment]
Thanks for the review! I addressed all comments now. I found that GatewayAllocator has some heuristic built in that checks how much of a shard on a node coincides with the primary before assigning replicas and added a check for sync ids there as well. Let me know if that makes sense.

File: src/main/java/org/elasticsearch/index/store/Store.java [hunk 15]
```diff
@@ -972,6 +976,11 @@ public void readFrom(StreamInput in) throws IOException {
             }
             this.metadata = builder.build();
             assert metadata.isEmpty() || numSegmentFiles() == 1 : "numSegmentFiles: " + numSegmentFiles();
+            commitUserData = new HashMap<>();
```
Comment (s1monw) [inline src/main/java/org/elasticsearch/index/store/Store.java]: can this be immutable?

File: src/main/java/org/elasticsearch/index/store/Store.java [hunk 16]
```diff
@@ -665,24 +664,29 @@ public String toString() {
...
 
         MetadataSnapshot() {
-            this.metadata = Collections.emptyMap();
+            this.metadata = Collections.EMPTY_MAP;
+            this.commitUserData = Collections.EMPTY_MAP;
         }
 
         MetadataSnapshot(IndexCommit commit, Directory directory, ESLogger logger) throws IOException {
-            metadata = buildMetadata(commit, directory, logger);
+            loadMetadata(commit, directory, logger);
```
Comment (s1monw) [inline src/main/java/org/elasticsearch/index/store/Store.java]: any chance we can return a tuple and make the two maps immutable?

File: src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java [hunk 17]
```diff
@@ -168,28 +169,43 @@ public void phase1(final SnapshotIndexCommit snapshot) throws ElasticsearchExcep
...
             // the source node
-            final Store.RecoveryDiff diff = recoverySourceMetadata.recoveryDiff(new Store.MetadataSnapshot(request.existingFiles()));
-            for (StoreFileMetaData md : diff.identical) {
-                response.phase1ExistingFileNames.add(md.name());
-                response.phase1ExistingFileSizes.add(md.length());
-                existingTotalSize += md.length();
-                if (logger.isTraceEnabled()) {
-                    logger.trace("[{}][{}] recovery [phase1] to {}: not recovering [{}], exists in local store and has checksum [{}], size [{}]",
-                            indexName, shardId, request.targetNode(), md.name(), md.checksum(), md.length());
+            final boolean recoverWithSyncId = recoverySourceMetadata.getCommitUserData().get(Engine.SYNC_COMMIT_ID) != null &&
```
Comment (s1monw) [inline src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: can you assign `recoverySourceMetadata.getCommitUserData().get(Engine.SYNC_COMMIT_ID)` to a local var makes it much easier to read

Comment (s1monw) [inline src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: `request.metadataSnapshot().getCommitUserData()` this one as well please

File: src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java [hunk 18]
```diff
@@ -155,12 +156,12 @@ private void doRecovery(final RecoveryStatus recoveryStatus) {
         assert recoveryStatus.sourceNode() != null : "can't do a recovery without a source node";
 
         logger.trace("collecting local files for {}", recoveryStatus);
-        Map<String, StoreFileMetaData> existingFiles;
```
Comment (s1monw) [inline src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java]: good :)

File: src/main/java/org/elasticsearch/indices/recovery/StartRecoveryRequest.java [hunk 19]
```diff
@@ -109,13 +109,9 @@ public void readFrom(StreamInput in) throws IOException {
         sourceNode = DiscoveryNode.readNode(in);
         targetNode = DiscoveryNode.readNode(in);
         markAsRelocated = in.readBoolean();
-        int size = in.readVInt();
-        existingFiles = Maps.newHashMapWithExpectedSize(size);
-        for (int i = 0; i < size; i++) {
-            StoreFileMetaData md = StoreFileMetaData.readStoreFileMetaData(in);
-            existingFiles.put(md.name(), md);
-        }
+        metadataSnapshot = Store.MetadataSnapshot.read(in);
```
Comment (s1monw) [inline src/main/java/org/elasticsearch/indices/recovery/StartRecoveryRequest.java]: good!!

File: src/main/java/org/elasticsearch/indices/store/TransportNodesListShardStoreMetaData.java [hunk 20]
```diff
@@ -216,15 +218,15 @@ public ShardId shardId() {
 
         @Override
         public Iterator<StoreFileMetaData> iterator() {
-            return files.values().iterator();
+            return metadataSnapshot.asMap().values().iterator();
```
Comment (s1monw) [inline src/main/java/org/elasticsearch/indices/store/TransportNodesListShardStoreMetaData.java]: you can just return `metadataSnapshot` it impls iterable

File: src/main/java/org/elasticsearch/indices/store/TransportNodesListShardStoreMetaData.java [hunk 21]
```diff
@@ -237,22 +239,18 @@ public static StoreFilesMetaData readStoreFilesMetaData(StreamInput in) throws I
...
             shardId.writeTo(out);
-            out.writeVInt(files.size());
-            for (StoreFileMetaData md : files.values()) {
-                md.writeTo(out);
-            }
+            metadataSnapshot.writeTo(out);
+        }
+
+        public String syncId() {
+            return metadataSnapshot.getCommitUserData().get(Engine.SYNC_COMMIT_ID);
```
Comment (s1monw) [inline src/main/java/org/elasticsearch/indices/store/TransportNodesListShardStoreMetaData.java]: maybe we add a utiltiy method to metadata snapshot ?

**Comment by s1monw** [comment]
this looks very good I left some minor comments

**Comment by s1monw** [comment]
> I found that GatewayAllocator has some heuristic built in that checks how much of a shard on a node coincides with the primary before assigning replicas and added a check for sync ids there as well. Let me know if that makes sense.

it does make a lot of sense to me

File: src/main/java/org/elasticsearch/index/store/Store.java (hunk 16 above)
Comment (brwe) [inline src/main/java/org/elasticsearch/index/store/Store.java]: ok. also, made both metadata and commitUserData final and ImmutableMap

**Comment by brwe** [comment]
Thanks for the review! Addressed all comments.

File: src/main/java/org/elasticsearch/index/store/Store.java [hunk 22]
```diff
@@ -657,32 +657,42 @@ public String toString() {
      *
      * @see StoreFileMetaData
      */
-    public final static class MetadataSnapshot implements Iterable<StoreFileMetaData>, Streamable {
+    public final static class MetadataSnapshot implements Iterable<StoreFileMetaData> {
```
Comment (s1monw) [inline src/main/java/org/elasticsearch/index/store/Store.java]: we have a new interface `Writeable.java` that we can use here. It has:

```Java
    void writeTo(StreamOutput out) throws IOException;
    T readFrom(StreamInput in) throws IOException;
```
and it's usually implemented by passing the `StreamInput` to a dedicated ctor so everything can be final. Can we use that here too?

**Comment by s1monw** [comment]
left one comment other than that LGTM

**Comment by clintongormley** [comment]
@brwe can you add a description to the PR plus a link to the main issue?

**Comment by brwe** [comment]
Pushed another commit.

File: src/main/java/org/elasticsearch/index/store/Store.java [hunk 23]
```diff
@@ -1010,6 +1032,15 @@ private final int numSegmentFiles() { // only for asserts
             }
             return count;
         }
+
```
Comment (s1monw) [inline src/main/java/org/elasticsearch/index/store/Store.java]: java docs?

File: src/main/java/org/elasticsearch/indices/recovery/RecoveryCleanFilesRequest.java [hunk 24]
```diff
@@ -61,7 +61,7 @@ public void readFrom(StreamInput in) throws IOException {
         super.readFrom(in);
         recoveryId = in.readLong();
         shardId = ShardId.readShardId(in);
-        snapshotFiles = Store.MetadataSnapshot.read(in);
+        snapshotFiles = new Store.MetadataSnapshot(in);
```
Comment (s1monw) [inline src/main/java/org/elasticsearch/indices/recovery/RecoveryCleanFilesRequest.java]: good

**Comment by s1monw** [comment]
it looks good to me to go on the feature branch... I left two minor comments.

**Comment by brwe** [comment]
added another commit with docs

**Comment by s1monw** [comment]
++ go push

**Comment by brwe** [comment]
pushed to https://github.com/elastic/elasticsearch/tree/feature/synced_flush

**Pull request closed without merging**